src/main/java/com/conveyal/gtfs/loader/Table.java

```diff
@@ -746,37 +747,38 @@ public void checkReferencesAndUniqueness(String keyValue, int lineNumber, Field
             }
         }
 
+        // These hold references to the set of IDs to check for duplicates and the ID to check. These depend on
+        // whether an order field is part of the "unique ID."
+        Set<String> listOfUniqueIds = referenceTracker.transitIds;
+        String uniqueId = transitId;
+
+        // There is no need to check for duplicate IDs if the field is a foreign reference.
+        if (field.isForeignReference()) return;
+        // Next, check that the ID is table-unique.
         if (field.name.equals(uniqueKeyField)) {
-            // These hold references to the set of IDs to check for duplicates and the ID to check. These depend on
-            // whether an order field is part of the "unique ID."
-            String transitId = String.join(":", getKeyFieldName(), keyValue);
-            Set<String> listOfUniqueIds;
-            String uniqueId;
             // Check for duplicate IDs and store entity-scoped IDs for referential integrity check
             if (isOrderField) {
                 // Check duplicate reference in set of field-scoped id:sequence (e.g., stop_sequence:12345:2)
                 // This should not be scoped by key field because there may be conflicts (e.g., with trip_id="12345:2"
                 listOfUniqueIds = referenceTracker.transitIdsWithSequence;
                 uniqueId = String.join(":", field.name, keyValue, string);
```

Comment:
This shortcut using string concatenation is only correct most of the time, and if the constituent strings never contain colons.  I think this might be what you are referring to by "should not be scoped by key field because there may be conflicts" because you cite a trip_id containing a colon character. We should probably use a Tuple type (which don't exist in Java but get defined in libraries and places in our own code). Or just a List of String. Though these things are going to get huge for stop_times tables.

Reply:
Can you name a case where this would not work?  Are you suggesting that the `listofUniqueIds` should be a List rather than a Set (also perhaps the variable is not well-named).

Reply:
Hmm, I guess it may always work since tables with an order field produce strings that are guaranteed to end in a colon followed by an integer. But I don't fully understand why these are being scoped by the name of the key field. Aren't all the identifiers coming from the same table part of the same namespace? Why would we be checking strings from different tables within a single reference tracker? I will need to get into this more deeply to understand how it works.

Reply:
Oh, I see you're checking cross-table references and uniqueness of keys all at the same time. I get the sense this could be refactored to be much more readable. There are also terms used in the comments like `transit ID` which I don't believe are defined anywhere. Maybe we can go over this together some time and try to clean it up. It does seem like it would properly perform the needed checks, but we need to improve the maintainability/readability of this section.